File: NimdaSecurity/Greedy/Java/BOJ14469.java
import java.lang.reflect.Array;
import java.util.*;
import java.util.Scanner;

public class Main {

    static class Pair implements Comparable<Pair>{

        int first;
        int second;

        public Pair(int first, int second) {
            this.first = first;
            this.second = second;
        }

        // 첫 번째 value (first) 기준 오름차순
        @Override
        public int compareTo(Pair p) {
            return Integer.compare(this.first, p.first);
        }
    }

    public static void main(String[] args) {

        Scanner input = new Scanner(System.in);
        int n = input.nextInt();
        ArrayList<Pair> cow_info = new ArrayList<>();
        int startTime, endTime;
        for (int i = 0; i < n; i++) {
            startTime = input.nextInt();
            endTime = input.nextInt();
            cow_info.add(new Pair(startTime, endTime));
        }

        Collections.sort(cow_info);
        int current_time = 0; // 현재 시간
        int progress_time = 0; // 진행 시간
        for (int i = 0; i < n; i++) {
            int waiting_time = 0; // 대기시간
            if (current_time < cow_info.get(i).first) {
                waiting_time = cow_info.get(i).first - current_time;
            }
            progress_time = cow_info.get(i).second;
            current_time += waiting_time + progress_time;
        }

        System.out.println(current_time);

        // 2 1 5 7 8 3

    }

}

File: NimdaSecurity/Greedy/C++/BOJ14469.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;


int main() {

    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    int n = 0;
    cin >> n;
    vector<pair<int, int>> cowInfo;
    int startTime, endTime;
    for (int i = 0; i < n; i++) {
        cin >> startTime >> endTime;
        cowInfo.push_back({startTime, endTime});
    }
    sort(cowInfo.begin(), cowInfo.end());

    int currentTime = 0; // 현재 시간
    int progress_time = 0; // 진행시간
    for (int i = 0; i < n; i++) {
        int waiting_time = 0; // 대기시간
        if (currentTime < cowInfo[i].first) {
            waiting_time = cowInfo[i].first - currentTime;
        }
        progress_time = cowInfo[i].second;
        currentTime += waiting_time + progress_time;
    }
    cout << currentTime << endl;

    return 0;
}

// N마리의 소가 농장에 방문한다.
// 소가 도착한 시간, 검문받는 시간은 상이하다. (같을 수도 있다.)
// 두 소가 동시에 검문을 받는것은 불가능하다.
